Bug: rake shakapacker:migrate_to_swc creates conflicting SWC configuration

# Bug: `rake shakapacker:migrate_to_swc` creates conflicting SWC configuration

## Environment
- Shakapacker version: 9.0.0.beta.8
- SWC version: @swc/core@^1.3.100

## Description

The `rake shakapacker:migrate_to_swc` task generates a `.swcrc` file that conflicts with Shakapacker's own default swc-loader configuration, causing builds to fail with:

**Note: This affects both JavaScript and TypeScript projects** - the swc-loader applies the same `env` config regardless of file type.

```
ERROR: `env` and `jsc.target` cannot be used together
```

## Root Cause

### 1. The Migration Task Creates (lib/shakapacker/swc_migrator.rb:33-50)
```ruby
DEFAULT_SWCRC_CONFIG = {
  "jsc" => {
    "parser" => {
      "syntax" => "ecmascript",
      "jsx" => true,
      "dynamicImport" => true
    },
    "transform" => {
      "react" => {
        "runtime" => "automatic"
      }
    },
    "target" => "es2015"  # ← This causes the conflict
  },
  "module" => {
    "type" => "es6"
  }
}.freeze
```

### 2. But Shakapacker's SWC Loader Adds (package/swc/index.ts:43-47)
```javascript
const defaultConfig: RuleSetRule = {
  loader: require.resolve("swc-loader"),
  options: {
    jsc: {
      parser: { /* ... */ },
      loose: true
    },
    sourceMaps: true,
    env: {           // ← This conflicts with jsc.target
      coreJs: 3,
      exclude: ["transform-typeof-symbol"],
      mode: "entry"
    }
  }
}
```

### 3. SWC Rejects the Merged Config

When webpack merges the `.swcrc` with the loader config, SWC receives both `jsc.target` and `env`, which are mutually exclusive as of @swc/core@1.3.72+.

## Steps to Reproduce

1. Create a new Rails app with Shakapacker 9.0.0.beta.8 configured with Babel
2. Run `rake shakapacker:migrate_to_swc`
3. Run `bin/shakapacker` or start the dev server
4. Observe build failure: `ERROR: \`env\` and \`jsc.target\` cannot be used together`

## Expected Behavior

The migration should create a `.swcrc` that works with Shakapacker's default loader configuration without conflicts.

## Proposed Fix

**Option 1: Remove `jsc.target` from the migrator** (Recommended)

Since Shakapacker's loader already provides `env` configuration with browserslist-style targeting, the `.swcrc` should not include `jsc.target`:

```ruby
DEFAULT_SWCRC_CONFIG = {
  "jsc" => {
    "parser" => {
      "syntax" => "ecmascript",
      "jsx" => true,
      "dynamicImport" => true
    },
    "transform" => {
      "react" => {
        "runtime" => "automatic"
      }
    }
    # Remove "target" => "es2015"
  },
  "module" => {
    "type" => "es6"
  }
}.freeze
```

**Option 2: Remove `env` from the loader's default config**

If you prefer to use `jsc.target` for simplicity, remove the `env` config from `package/swc/index.ts`. However, this loses the browserslist integration and fine-grained control.

**Option 3: Make it configurable**

Allow users to choose between `jsc.target` and `env` via a configuration option, but this adds complexity.

## Additional Issue: ESLint Dependencies

The `rake shakapacker:clean_babel_packages` task removes `@babel/core` and `@babel/eslint-parser` from package.json. However, many projects use ESLint configurations (like `eslint-config-shakacode`, `eslint-config-airbnb`, etc.) that depend on `@babel/eslint-parser`.

After running the clean task, ESLint fails with:
```
Error: Cannot find module '@babel/eslint-parser'
```

### Suggested Solutions

1. **Documentation**: Add a note in the migration guide that `@babel/core` and `@babel/eslint-parser` should be kept if using Babel-dependent ESLint configs (TypeScript projects using `@typescript-eslint/parser` are unaffected)
2. **Exclude from removal**: Don't remove these packages automatically, or prompt the user
3. **Update BABEL_PACKAGES list**: Add a comment explaining these are dev-only linting dependencies

**Note**: TypeScript projects using `@typescript-eslint/parser` instead of `@babel/eslint-parser` don't have this issue.

## Workaround (Current)

Manually edit the generated `.swcrc` to remove `"target": "es2015"` and keep `@babel/core` and `@babel/eslint-parser` as devDependencies:

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": true,
      "dynamicImport": true
    },
    "transform": {
      "react": {
        "runtime": "automatic"
      }
    }
  },
  "module": {
    "type": "es6"
  }
}
```

```json
// package.json devDependencies
{
  "@babel/core": "^7.20.0",
  "@babel/eslint-parser": "^7.16.5"
}
```

## References

- SWC documentation on target vs env: https://swc.rs/docs/configuration/compilation
- Related SWC issue discussing the conflict: https://github.com/swc-project/swc/issues/5815

## Test Case Repository

You can see the working configuration (after manual fixes) in this PR:
https://github.com/shakacode/react-webpack-rails-tutorial/pull/669

--- a/lib/shakapacker/doctor.rb
+++ b/lib/shakapacker/doctor.rb
@@ -432,6 +432,28 @@ module Shakapacker
         if transpiler == "esbuild" && package_installed?("babel-loader")
           @warnings << "Both esbuild and Babel dependencies are installed. Consider removing Babel dependencies to reduce node_modules size"
         end
+
+        # Check for SWC configuration conflicts
+        if transpiler == "swc"
+          check_swc_config_conflicts
+        end
+      end
+
+      def check_swc_config_conflicts
+        swcrc_path = root_path.join(".swcrc")
+        return unless swcrc_path.exist?
+
+        begin
+          swcrc = JSON.parse(File.read(swcrc_path))
+          # Check for conflicting jsc.target and env settings
+          if swcrc.dig("jsc", "target") && swcrc["env"]
+            @issues << "SWC configuration conflict: .swcrc contains both 'jsc.target' and 'env' settings, which are mutually exclusive. Remove 'jsc.target' from .swcrc"
+          elsif swcrc.dig("jsc", "target")
+            @warnings << "SWC configuration: .swcrc contains 'jsc.target' which may conflict with the loader's 'env' setting. Consider removing 'jsc.target' from .swcrc to avoid build errors"
+          end
+        rescue JSON::ParserError
+          @warnings << "SWC configuration: .swcrc exists but contains invalid JSON"
+        end
       end
 
       def check_css_dependencies

--- a/lib/shakapacker/swc_migrator.rb
+++ b/lib/shakapacker/swc_migrator.rb
@@ -41,8 +41,7 @@ module Shakapacker
           "react" => {
             "runtime" => "automatic"
           }
-        },
-        "target" => "es2015"
+        }
       },
       "module" => {
         "type" => "es6"
@@ -67,6 +66,10 @@ module Shakapacker
       logger.info "🎉 Migration to SWC complete!"
       logger.info "   Note: SWC is approximately 20x faster than Babel for transpilation."
       logger.info "   Please test your application thoroughly after migration."
+      logger.info "\n📝 Configuration Info:"
+      logger.info "   - .swcrc provides base configuration for all environments"
+      logger.info "   - The SWC loader adds automatic environment targeting (via 'env' setting)"
+      logger.info "   - You can customize .swcrc, but avoid setting 'jsc.target' as it conflicts with 'env'"
 
       # Show cleanup recommendations if babel packages found
       if results[:babel_packages_found].any?
@@ -97,6 +100,16 @@ module Shakapacker
       unless package_json_path.exist?
         logger.error "❌ No package.json found"
         return { removed_packages: [], config_files_deleted: [] }
+      end
+
+      # Check if ESLint uses Babel parser
+      if eslint_uses_babel?
+        logger.info "\n⚠️  WARNING: ESLint configuration detected that may use Babel"
+        logger.info "   If you use @babel/eslint-parser or babel-eslint, you may need to:"
+        logger.info "   1. Keep @babel/core and related Babel packages for ESLint"
+        logger.info "   2. Or switch to @typescript-eslint/parser for TypeScript files"
+        logger.info "   3. Or use espree (ESLint's default parser) for JavaScript files"
+        logger.info "\n   Proceeding with Babel package removal. Check your ESLint config after."
       end
 
       removed_packages = remove_babel_from_package_json(package_json_path)
@@ -131,6 +144,48 @@ module Shakapacker
     end
 
     private
+
+      def eslint_uses_babel?
+        # Check for ESLint config files
+        eslint_configs = [
+          ".eslintrc",
+          ".eslintrc.js",
+          ".eslintrc.cjs",
+          ".eslintrc.yaml",
+          ".eslintrc.yml",
+          ".eslintrc.json"
+        ]
+
+        eslint_configs.each do |config_file|
+          config_path = root_path.join(config_file)
+          next unless config_path.exist?
+
+          content = File.read(config_path)
+          # Check for Babel parser references
+          return true if content.match?(/@babel\/eslint-parser|babel-eslint/)
+        end
+
+        # Check package.json for eslintConfig
+        package_json_path = root_path.join("package.json")
+        if package_json_path.exist?
+          begin
+            package_json = JSON.parse(File.read(package_json_path))
+            if package_json["eslintConfig"]
+              return true if package_json["eslintConfig"].to_s.match?(/@babel\/eslint-parser|babel-eslint/)
+            end
+
+            # Check if Babel ESLint packages are installed
+            dependencies = package_json["dependencies"] || {}
+            dev_dependencies = package_json["devDependencies"] || {}
+            all_deps = dependencies.merge(dev_dependencies)
+            return true if all_deps.key?("@babel/eslint-parser") || all_deps.key?("babel-eslint")
+          rescue JSON::ParserError
+            # Ignore parse errors
+          end
+        end
+
+        false
+      end
 
       def update_shakapacker_config
         config_path = root_path.join("config/shakapacker.yml")

--- a/lib/shakapacker/swc_migrator.rb
+++ b/lib/shakapacker/swc_migrator.rb
@@ -171,7 +171,7 @@ module Shakapacker
           begin
             package_json = JSON.parse(File.read(package_json_path))
             if package_json["eslintConfig"]
-              return true if package_json["eslintConfig"].to_s.match?(/@babel\/eslint-parser|babel-eslint/)
+              return true if package_json["eslintConfig"].to_json.match?(/@babel\/eslint-parser|babel-eslint/)
             end
 
             # Check if Babel ESLint packages are installed

--- a/lib/shakapacker/swc_migrator.rb
+++ b/lib/shakapacker/swc_migrator.rb
@@ -30,6 +30,15 @@ module Shakapacker
       "swc-loader" => "^0.2.6"
     }.freeze
 
+    ESLINT_CONFIG_FILES = %w[
+      .eslintrc
+      .eslintrc.js
+      .eslintrc.cjs
+      .eslintrc.yaml
+      .eslintrc.yml
+      .eslintrc.json
+    ].freeze
+
     DEFAULT_SWCRC_CONFIG = {
       "jsc" => {
         "parser" => {
@@ -147,16 +156,10 @@ module Shakapacker
 
       def eslint_uses_babel?
         # Check for ESLint config files
-        eslint_configs = [
-          ".eslintrc",
-          ".eslintrc.js",
-          ".eslintrc.cjs",
-          ".eslintrc.yaml",
-          ".eslintrc.yml",
-          ".eslintrc.json"
-        ]
-
-        eslint_configs.each do |config_file|
+        # Note: This is a heuristic check that may have false positives (e.g., in comments),
+        # but false positives only result in an extra warning, which is safer than silently
+        # breaking ESLint configurations.
+        ESLINT_CONFIG_FILES.each do |config_file|
           config_path = root_path.join(config_file)
           next unless config_path.exist?
 
@@ -179,8 +182,8 @@ module Shakapacker
             dev_dependencies = package_json["devDependencies"] || {}
             all_deps = dependencies.merge(dev_dependencies)
             return true if all_deps.key?("@babel/eslint-parser") || all_deps.key?("babel-eslint")
-          rescue JSON::ParserError
-            # Ignore parse errors
+          rescue JSON::ParserError => e
+            logger.debug "Could not parse package.json for ESLint detection: #{e.message}"
           end
         end
 

--- a/spec/shakapacker/doctor_spec.rb
+++ b/spec/shakapacker/doctor_spec.rb
@@ -853,6 +853,56 @@ describe Shakapacker::Doctor do
           expect(doctor.warnings).to include(match(/Both SWC and Babel dependencies are installed/))
         end
       end
+
+      context "with SWC config containing jsc.target" do
+        before do
+          package_json = {
+            "devDependencies" => {
+              "@swc/core" => "^1.3.0",
+              "swc-loader" => "^0.2.0"
+            }
+          }
+          File.write(package_json_path, JSON.generate(package_json))
+          File.write(root_path.join(".swcrc"), JSON.generate({
+            "jsc" => {
+              "target" => "es2015",
+              "parser" => {
+                "syntax" => "ecmascript"
+              }
+            }
+          }))
+        end
+
+        it "warns about potential config conflict" do
+          doctor.send(:check_javascript_transpiler_dependencies)
+          expect(doctor.warnings).to include(match(/SWC configuration.*jsc\.target.*may conflict/))
+        end
+      end
+
+      context "with SWC config containing both jsc.target and env" do
+        before do
+          package_json = {
+            "devDependencies" => {
+              "@swc/core" => "^1.3.0",
+              "swc-loader" => "^0.2.0"
+            }
+          }
+          File.write(package_json_path, JSON.generate(package_json))
+          File.write(root_path.join(".swcrc"), JSON.generate({
+            "jsc" => {
+              "target" => "es2015"
+            },
+            "env" => {
+              "targets" => "defaults"
+            }
+          }))
+        end
+
+        it "reports a configuration error" do
+          doctor.send(:check_javascript_transpiler_dependencies)
+          expect(doctor.issues).to include(match(/SWC configuration conflict.*both 'jsc\.target' and 'env'/))
+        end
+      end
     end
   end
 

--- a/spec/shakapacker/swc_migrator_spec.rb
+++ b/spec/shakapacker/swc_migrator_spec.rb
@@ -193,6 +193,27 @@ describe Shakapacker::SwcMigrator do
   end
 
   describe "#clean_babel_packages" do
+    context "with ESLint using Babel parser" do
+      before do
+        File.write(root_path.join("package.json"), JSON.pretty_generate({
+          "name": "test-app",
+          "devDependencies": {
+            "@babel/core": "^7.20.0",
+            "@babel/eslint-parser": "^7.20.0",
+            "babel-loader": "^9.1.0"
+          }
+        }))
+        File.write(root_path.join(".eslintrc.js"), 'module.exports = { parser: "@babel/eslint-parser" }')
+      end
+
+      it "warns about ESLint using Babel" do
+        migrator.clean_babel_packages(run_installer: false)
+
+        expect(logger).to have_received(:info).with(/WARNING: ESLint configuration detected/)
+        expect(logger).to have_received(:info).with(/switch to @typescript-eslint\/parser/)
+      end
+    end
+
     context "with babel packages installed" do
       before do
         File.write(root_path.join("package.json"), JSON.pretty_generate({
@@ -372,7 +393,7 @@ describe Shakapacker::SwcMigrator do
     it "defines default SWC configuration" do
       config = described_class::DEFAULT_SWCRC_CONFIG
       expect(config["jsc"]["parser"]["jsx"]).to eq(true)
-      expect(config["jsc"]["target"]).to eq("es2015")
+      expect(config["jsc"]["target"]).to be_nil
       expect(config["module"]["type"]).to eq("es6")
     end
   end

--- a/spec/shakapacker/swc_migrator_spec.rb
+++ b/spec/shakapacker/swc_migrator_spec.rb
@@ -193,7 +193,7 @@ describe Shakapacker::SwcMigrator do
   end
 
   describe "#clean_babel_packages" do
-    context "with ESLint using Babel parser" do
+    context "with ESLint using Babel parser in .eslintrc" do
       before do
         File.write(root_path.join("package.json"), JSON.pretty_generate({
           "name": "test-app",
@@ -207,6 +207,29 @@ describe Shakapacker::SwcMigrator do
       end
 
       it "warns about ESLint using Babel" do
+        migrator.clean_babel_packages(run_installer: false)
+
+        expect(logger).to have_received(:info).with(/WARNING: ESLint configuration detected/)
+        expect(logger).to have_received(:info).with(/switch to @typescript-eslint\/parser/)
+      end
+    end
+
+    context "with ESLint using Babel parser in package.json" do
+      before do
+        File.write(root_path.join("package.json"), JSON.pretty_generate({
+          "name": "test-app",
+          "devDependencies": {
+            "@babel/core": "^7.20.0",
+            "babel-loader": "^9.1.0"
+          },
+          "eslintConfig": {
+            "parser": "@babel/eslint-parser",
+            "extends": ["eslint:recommended"]
+          }
+        }))
+      end
+
+      it "warns about ESLint using Babel in package.json" do
         migrator.clean_babel_packages(run_installer: false)
 
         expect(logger).to have_received(:info).with(/WARNING: ESLint configuration detected/)

